external/js/cothority/src/network/connection.ts

```diff
@@ -132,42 +131,85 @@ export class WebSocketConnection implements IConnection {
 /**
  * Multi peer connection that tries all nodes one after another
  */
-export class RosterWSConnection extends WebSocketConnection {
+export class RosterWSConnection {
+    static parallel: number = 2;
     addresses: string[];
+    addressNext: number;
+    connections: WebSocketConnection[] = [];
 
     /**
      * @param r         The roster to use
      * @param service   The name of the service to reach
-     * @param retry = 2 How many times to retry a failing connection with another node
      */
-    constructor(r: Roster, service: string, private retry: number = 2) {
-        super("", service);
+    constructor(r: Roster, service: string) {
         this.addresses = r.list.map((conode) => conode.getWebSocketAddress());
         shuffle(this.addresses);
+        this.addressNext = RosterWSConnection.parallel;
+        if (this.addressNext > this.addresses.length) {
+            this.addressNext = this.addresses.length;
+        }
+        for (let i = 0; i < this.addressNext; i++) {
+            this.connections.push(new WebSocketConnection(this.addresses[i], service));
```

Comment:
Please create an array of addresses instead so you don't have to have the URL modifiable. A new ws is created anyway when sending so creating the connection object is cheap.

Reply:
That complicates things a lot if I want to keep some memory of the nodes that are accessible and the ones that are not accessible. With the current solution, if a node fails, it won't be used until it rotates back in. If I put all nodes in a list, then I need to re-arrange the list to make sure the failing node is at the end. Possible but less elegant, IMHO.

Also, if there are many nodes (100s), we'll keep them all around for no apparent reason. Is that not a problem?

Reply:
No you don't need to change a lot. Just remove the `connections` array and create them in the `do ... while`, I'm not asking more. You already have the array of addresses (sorry that wasn't clear).

Reply:
Hmm - I don't see how that would work. The current `connections` is kind of a pool of workers. Every time I want to send a command to the nodes, I start all workers until one succeeds. While waiting for the first success, every failing worker will change its address and try again.

This means that if there are failing nodes, the workers in `connections` have their addresses changed so as not to include those failing nodes. A subsequent call to `send` then only tries nodes that didn't fail before.

I'm feeling I'm missing again something, sorry ;)

Reply:
You can hardly talk about workers because on the `send` it will create a new websocket every time. That's why I was saying it doesn't change much if you create a connection for each address, the end result will be the same in terms of memory footprint (almost). The small overhead allows you to keep the URL unique per connection which was my first goal with the object (allowing later optimization to reuse connections). Thus I was proposing something like:

```javascript
let promises = [];
for (let i = 0; i < 2; i++) {
  promises.push(new Promise((r, rr) => {
    let connection = new WebSocketConnection(this.addresses[i], this.service);
    // do ... while
  });
}
return Promise.race(promises);
```

Reply:
But like this you always send the requests to the same conodes. So suppose `parallel = 2` and the first two conodes in the list are down. This means that every time you send a request, you have to wait for those conodes to time out before you can try the next ones...